### PR TITLE
chore: Update datafusion

### PR DESCRIPTION
**Pull request by alamb**
# Rationale:
1. Keep up with changes upstream so it remains easy to pull in the latest datafusion
1. Notable PRs include https://github.com/apache/arrow-datafusion/pull/1315, https://github.com/apache/arrow-datafusion/pull/1319 and https://github.com/apache/arrow-datafusion/pull/1163

# Changes:
1. Update to use latest datafusion: https://github.com/apache/arrow-datafusion/commit/66f5d1906992dcf4f5b7ab787a39a4105c2f5b42
2. Update query for `LogicalPlan` struct extraction in https://github.com/apache/arrow-datafusion/issues/1228

### CODE REVIEW

File: query_tests/cases/in/duplicates.expected [hunk 1]
```diff
@@ -3,10 +3,10 @@
 +---------------+---------------------------------------------------------------------------------------------------------------------------------------------+
 | plan_type     | plan                                                                                                                                        |
 +---------------+---------------------------------------------------------------------------------------------------------------------------------------------+
-| logical_plan  | Sort: #h2o.time ASC NULLS FIRST, #h2o.state ASC NULLS FIRST, #h2o.city ASC NULLS FIRST                                                      |
+| logical_plan  | Sort: #h2o.time ASC NULLS LAST, #h2o.state ASC NULLS LAST, #h2o.city ASC NULLS LAST                                                         |
```
Comment (alamb) [inline query_tests/cases/in/duplicates.expected:6]: due to https://github.com/apache/arrow-datafusion/pull/1344

File: query_tests/cases/in/no_stats_plans.expected [hunk 2]
```diff
@@ -8,7 +8,7 @@
 |               |     TableScan: h2o projection=Some([0])                     |
 | physical_plan | ProjectionExec: expr=[COUNT(UInt8(1))@0 as COUNT(UInt8(1))] |
 |               |   RepartitionExec: partitioning=RoundRobinBatch(4)          |
-|               |     ProjectionExec: expr=[3 as COUNT(Uint8(1))]             |
+|               |     ProjectionExec: expr=[3 as COUNT(UInt8(1))]             |
```
Comment (alamb) [inline query_tests/cases/in/no_stats_plans.expected:11]: This was a typo that got fixed in https://github.com/apache/arrow-datafusion/pull/1063

File: query_tests/cases/in/pushdown.expected [hunk 3]
```diff
@@ -165,16 +165,16 @@
...
+| logical_plan  | Projection: #restaurant.count, #restaurant.system, #restaurant.time, #restaurant.town                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
+|               |   Filter: Float64(5) < #restaurant.system AND #restaurant.town != Utf8("tewsbury") AND #restaurant.system < Float64(7) AND #restaurant.count = Int64(632) OR #restaurant.town = Utf8("reading") AND #restaurant.time > TimestampNanosecond(130) AS Float64(5) < restaurant.system AND restaurant.town != Utf8("tewsbury") AND restaurant.system < Float64(7) AND restaurant.count = Int64(632) OR restaurant.town = Utf8("reading") AND restaurant.time > totimestamp(Utf8("1970-01-01T00:00:00.000000130+00:00"))                                                          |
```
Comment (alamb) [inline query_tests/cases/in/pushdown.expected:172]: There is a new alias here due to https://github.com/apache/arrow-datafusion/pull/1319 (fixed by @viirya 👋 ) as the alias ensures the output matches the output column name

File: query/src/exec.rs [hunk 4]
```diff
@@ -140,7 +140,7 @@ impl Drop for Executor {
 pub fn make_schema_pivot(input: LogicalPlan) -> LogicalPlan {
     let node = Arc::new(SchemaPivotNode::new(input));
 
-    LogicalPlan::Extension { node }
+    LogicalPlan::Extension(ExtensionPlan { node })
```
Comment (xudong963) [inline query/src/exec.rs:143]: As I know, @liukun4515 will do a final refactoring which may rename `ExtensionPlan` to `Extension` and other changes to keep code style consistent. So I think you can do this ticket after final refactoring.

**Review by crepererum — APPROVED** [review]

File: query/src/exec.rs (hunk 4 above)
Comment (alamb) [reply query/src/exec.rs:143]: Thanks for the heads up @xudong963 -- I don't mind updating this code more than once to keep up with the latest DataFusion code. It is part of the cost of living on the bleeding edge :)

**Pull request merged**